applications/esl/peripheral_esl/hw/display/epd_shield_display.c

```diff
@@ -0,0 +1,259 @@
+		LOG_ERR("read image idx %d failed (err %d)", img_idx, err);
+	}
+
+	img_head = (struct waveshare_gray_head *)esl_obj->img_obj_buf;
+	buf_desc.width = img_head->w;
+	buf_desc.height = img_head->h;
+	buf_desc.pitch = buf_desc.width;
+	buf_desc.buf_size = (img_head->w * img_head->h) / 8;
+	img_size -= sizeof(struct waveshare_gray_head);
+
+	display_blanking_on(display_dev);
+	cur_pos = sizeof(struct waveshare_gray_head);
+	chunk_size = img_head->w;
+
+	err = esl_obj->cb.read_img_from_storage(img_idx, esl_obj->img_obj_buf, img_size, 0);
+	rc = display_write(display_dev, 0, 0, &buf_desc,
+			   (esl_obj->img_obj_buf + sizeof(struct display_buffer_descriptor)));
+	if (rc) {
+		LOG_ERR("display_write (rc %d)", rc);
+	}
+
+	display_blanking_off(display_dev);
+	LOG_DBG("Use Raw display interface API");
+#if defined(ESL_POWER_PROFILE)
+	(void)pm_device_action_run(spi, PM_DEVICE_ACTION_SUSPEND);
+	display_gpio_onoff(false);
+#endif
+	return 0;
+}
+
+void display_unassociated(uint8_t disp_idx)
+{
+
+#if defined(ESL_POWER_PROFILE)
+	display_gpio_onoff(true);
+	/**
+	 * To optimize power remove static declaration of ssd16xx_init in
+	 * zephyr/drivers/display/ssd16xx.c
+	 **/
+	ssd16xx_init(display_dev);
+	(void)pm_device_action_run(spi, PM_DEVICE_ACTION_RESUME);
+#endif
+#if (CONFIG_BT_ESL_PAINT_LIB)
+	struct bt_esls *esl_obj = esl_get_esl_obj();
+	struct display_buffer_descriptor buf_desc;
+	int err;
+
+	LOG_INF("capability x %d y %d", capabilities.x_resolution, capabilities.y_resolution);
+	buf_desc.width = capabilities.x_resolution;
+	buf_desc.height = capabilities.y_resolution;
+	buf_desc.pitch = capabilities.x_resolution;
+	buf_desc.buf_size = (buf_desc.width * buf_desc.height) / 8;
+
+	Paint_Init(&paint_black, esl_obj->img_obj_buf, capabilities.x_resolution,
+		   capabilities.y_resolution);
+
+	Paint_Clear(&paint_black, UNCOLORED);
+	Paint_SetRotate(&paint_black, 0);
+	Paint_DrawStringAt(&paint_black, 10, 10, "Hello Nordic!", &Font16, COLORED);
+	Paint_DrawStringAt(&paint_black, 10, 30, "UNAssociated", &Font16, COLORED);
+	Paint_DrawStringAt(&paint_black, 10, 50, "ESL TAG", &Font16, COLORED);
+	Paint_DrawStringAt(&paint_black, 10, 70, "APAC", &Font16, COLORED);
+	Paint_DrawStringAt(&paint_black, 10, 90, CONFIG_BT_DIS_MODEL, &Font16, COLORED);
+
+	display_blanking_on(display_dev);
+	err = display_write(display_dev, 0, 0, &buf_desc, esl_obj->img_obj_buf);
+	if (err) {
+		LOG_ERR("display_write (rc %d)", err);
+	}
+
+	display_blanking_off(display_dev);
+#endif /* CONFIG_BT_ESL_PAINT_LIB */
+#if defined(ESL_POWER_PROFILE)
+	(void)pm_device_action_run(spi, PM_DEVICE_ACTION_SUSPEND);
+#endif
+}
+
+void display_associated(uint8_t disp_idx)
+{
+
+#if defined(ESL_POWER_PROFILE)
+	display_gpio_onoff(true);
+	/**
+	 * To optimize power remove static declaration of ssd16xx_init in
+	 * zephyr/drivers/display/ssd16xx.c
+	 **/
+	ssd16xx_init(display_dev);
+	(void)pm_device_action_run(spi, PM_DEVICE_ACTION_RESUME);
+#endif
+#if (CONFIG_BT_ESL_PAINT_LIB)
```

Comment:
where is this lib from? I don't see it in zephyr or sdk-nrf

Reply:
Discussion offline.